mlfinlab/labeling/labeling.py

```diff
@@ -0,0 +1,288 @@
+        out.loc[loc, 'sl'] = df0[df0 < stop_loss[loc]].index.min()  # earliest stop loss
+        out.loc[loc, 'pt'] = df0[df0 > profit_taking[loc]].index.min()  # earliest profit taking
+
+    return out
+
+
+# Snippet 3.4 page 49, Adding a Vertical Barrier
+def add_vertical_barrier(t_events, close, num_days=1):
+    """
+    Snippet 3.4 page 49, Adding a Vertical Barrier
+
+    For each index in t_events, it finds the timestamp of the next price bar at or immediately after
+    a number of days num_days. This vertical barrier can be passed as an optional argument t1 in get_events.
+
+    This function creates a series that has all the timestamps of when the vertical barrier would be reached.
+
+    :param t_events: (series) series of events (symmetric CUSUM filter)
+    :param close: (series) close prices
+    :param num_days: (int) maximum number of days a trade can be active
+    :return: (series) timestamps of vertical barriers
+    """
+
+    vertical_barriers = close.index.searchsorted(t_events + pd.Timedelta(days=num_days))
+    vertical_barriers = vertical_barriers[vertical_barriers < close.shape[0]]
+    vertical_barriers = pd.Series(close.index[vertical_barriers],
+                                  index=t_events[:vertical_barriers.shape[0]])  # NaNs at end
+    return vertical_barriers
+
+
+# Snippet 3.3 -> 3.6 page 50, Getting the Time of the First Touch, with Meta Labels
+def get_events(close, t_events, pt_sl, target, min_ret, num_threads, vertical_barrier_times=False, side=None):
+    """
+    Snippet 3.6 page 50, Getting the Time of the First Touch, with Meta Labels
+
+    This function is orchestrator to meta-label the data, in conjunction with the Triple Barrier Method.
+
+    :param close: (series) Close prices
+    :param t_events: (series) of t_events. These are timestamps that will seed every triple barrier.
+        These are the timestamps selected by the sampling procedures discussed in Chapter 2, Section 2.5.
+        Eg: CUSUM Filter
+    :param pt_sl: (2 element array) element 0, indicates the profit taking level; element 1 is stop loss level.
+        A non-negative float that sets the width of the two barriers. A 0 value means that the respective
+        horizontal barrier (profit taking and/or stop loss) will be disabled.
+    :param target: (series) of values that are used (in conjunction with pt_sl) to determine the width
+        of the barrier. In this program this is daily volatility series.
+    :param min_ret: (float) The minimum target return required for running a triple barrier search.
+    :param num_threads: (int) The number of threads concurrently used by the function.
+    :param vertical_barrier_times: (series) A pandas series with the timestamps of the vertical barriers.
+        We pass a False when we want to disable vertical barriers.
+    :param side: (series) Side of the bet (long/short) as decided by the primary model
```

Comment:
Can we set it as 'model_side', or 'model_pred'?

Reply:
Changed to side_prediction. Good idea.